### PR TITLE
feat(lsp): map K to hover by default

### PR DESCRIPTION
**Pull request by mfussenegger**
Relates to https://github.com/neovim/neovim/issues/24252

### CODE REVIEW

File: runtime/lua/vim/lsp.lua [hunk 1]
```diff
@@ -1212,6 +1215,13 @@ function lsp.start_client(config)
     if vim.bo[bufnr].formatexpr == 'v:lua.vim.lsp.formatexpr()' then
       vim.bo[bufnr].formatexpr = nil
     end
+    local keymaps = api.nvim_buf_get_keymap(bufnr, "n")
```
Comment (folke) [inline runtime/lua/vim/lsp.lua:1218]: You may want to use `  local map = vim.fn.maparg("K", "n", false, true)` here instead which is about 200x faster than getting all keymaps

Comment (mfussenegger) [reply runtime/lua/vim/lsp.lua]: Thanks for pointing out `maparg`, wasn't aware of it.

It's btw. only a 2-4x difference :)

Comment (folke) [reply runtime/lua/vim/lsp.lua]: In my config, it was about 1ms for getting all of them and 0.006ms with maparg. Profiles it on running the code 100 times with luajit cache cleared. I do have a lot of keymaps though 🙂

**Review by justinmk — APPROVED** [review]
Let's move forward with this. Reminder: the plan from https://github.com/neovim/neovim/issues/24252 is that users can either opt-out or opt-in (haven't decided which, yet).

**Comment by gpanders** [comment]
AS I mentioned in https://github.com/neovim/neovim/issues/24252#issuecomment-1634393710 I'm not in favor of this as a default mapping. Specifically for C and C++ buffers, this overrides the (very useful) default `'keywordprg'`, since `is_empty_or_default` returns true in those cases.

If we were to change this to instead only map `K` if `'keywordprg'` is empty I would not object as much, though ultimately I'd prefer to use a different mapping altogether so that hover and `'keywordprg'` could co-exist.

**Comment by folke** [comment]
@gpanders but wouldn't this code only trigger when an actual lsp server is attached to the buffer? Making lsp hover the default in that case makes a lot of sense to me imo

**Comment by gpanders** [comment]
> @gpanders but wouldn't this code only trigger when an actual lsp server is attached to the buffer? Making lsp hover the default in that case makes a lot of sense to me imo

Yes, but even when using LSP one may wish to use the `'keywordprg'` behavior. For instance, in C and C++ this is set to `:Man` by default. Viewing a man page for a function and viewing the LSP hover information cover different use cases, so users should be able to do both, without the latter clobbering the former.

**Comment by justinmk** [comment]
> AS I mentioned in [#24252 (comment)](https://github.com/neovim/neovim/issues/24252#issuecomment-1634393710) I'm not in favor of this as a default mapping. Specifically for C and C++ buffers, this overrides the (very useful) default `'keywordprg'`, since `is_empty_or_default` returns true in those cases.

@gpanders can't you the mapping? We have to be willing to steamroll some of the C-specific traditions, if we want to make useful progress on this "defaults" topic. This is similar to https://github.com/neovim/neovim/pull/22500

**Comment by folke** [comment]
As an alternative, for clangd, a user can disable the hover capability to keep the default hover

**Comment by justinmk** [comment]
Merging this so it doesn't stall. If the :Man problem is bigger than I think we can revisit.

**Comment by gpanders** [comment]
>@gpanders can't you override the mapping?

Of course. It's not really about what I can do (I know my way around well enough to take care of myself). I'm advocating for a certain position, which I suspect is not exactly niche (though maybe I'll be proved wrong on that).

File: runtime/doc/lsp.txt [hunk 2]
```diff
@@ -58,13 +58,16 @@ options are not restored when the LSP client is stopped or detached.
 - 'formatexpr' is set to |vim.lsp.formatexpr()|, so you can format lines via
   |gq| if the language server supports it.
   - To opt out of this use |gw| instead of gq, or set 'formatexpr' on LspAttach.
+- |K| is mapped to |vim.lsp.buf.hover()| unless |'keywordprg'| is customized or
+  a custom keymap for `K` exists.
```
Comment (justinmk) [inline runtime/doc/lsp.txt:62]: Also documented in vim_diff.txt and `:help K-lsp-default` : https://github.com/neovim/neovim/pull/24351

**Comment by gpanders** [comment]
At the risk of beating a dead horse, another demonstration of why `'keywordprg'` and LSP hover are not the same.

Using Python's default `'keywordprg'` (pydoc), pressing `K` over the token `msgpack` in a file shows this:

image

LSP's hover shows this:

image

These are complementary features. Hover should not be blocking the use of `'keywordprg'` by default. Making users opt-in to useful features is antithetical to the design goals of Nvim.

I'll leave it alone after this. But I wanted to make my argument for posterity, if nothing else.

**Comment by justinmk** [comment]
I agree. But also, LSP hover is essentially a different default 'keywordprg'. We have to choose one, and it's somewhat arbitrary to choose from:

- "most languages have no 'keywordprg', so K just does nothing there, sorry, or...
- "most languages have no 'keywordprg', so only _for those_ languages we'll set K to LSP hover", or...
- "most languages have no 'keywordprg', so to use LSP hover you need this other mapping (perhaps `gK`)..."

The old shell-tool behavior of 'keywordprg' (man for C, python doc for python) just was never implemented for most cases. It's hard to accept that K should be mostly useless for all cases except C/python, when LSP provides coverage for all LSP-supported buffers.

A related, maybe silly, idea: allow mappings to "stack". When a new mapping is assigned, the old mapping is pushed to a stack. Users can see the stack of mappings via some standard prefix like `<C-\>`, so e.g. `<C-\>K` would show the stack of K mappings. That way, buffers can add stuff but old mappings don't get lost, only somewhat buried.

**Comment by kaddkaka** [comment]
How to remap K to old behavior?

**Pull request merged**